src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java

```diff
@@ -1046,37 +1053,60 @@ public String getParentResourceType(final String resourceType) {
     public boolean isResourceType(final Resource resource, final String resourceType) {
         boolean result = false;
         if ( resource != null && resourceType != null ) {
-             // Check if the resource is of the given type. This method first checks the
-             // resource type of the resource, then its super resource type and continues
-             //  to go up the resource super type hierarchy.
-             if (ResourceTypeUtil.areResourceTypesEqual(resourceType, resource.getResourceType(), factory.getSearchPath())) {
-                 result = true;
-             } else {
-                 Set<String> superTypesChecked = new HashSet<>();
-                 String superType = this.getParentResourceType(resource);
-                 while (!result && superType != null) {
-                     if (ResourceTypeUtil.areResourceTypesEqual(resourceType, superType, factory.getSearchPath())) {
-                         result = true;
-                     } else {
-                         superTypesChecked.add(superType);
-                         superType = this.getParentResourceType(superType);
-                         if (superType != null && superTypesChecked.contains(superType)) {
-                             throw new SlingException("Cyclic dependency for resourceSuperType hierarchy detected on resource " + resource.getPath(), null);
-                         }
-                     }
-                 }
+
+             // Check if the result is already available from cache
+             ResourceTypeInformation key = new ResourceTypeInformation(resource.getResourceType(),resource.getResourceSuperType(), resourceType);
+             Boolean value = resourceTypeLookupCache.get(key);
```

Comment:
Thinking about it some more, maybe a `ConcurrentMap` is a better fit? It's more expressive, e.g.

```java
resourceTypeLookupCache.computeIfAbset(key, (key) -> isResourceTypeInternal(resource, resourceType));
```

and also skips the locking of the synchronized map.

If you're happy with the synchronized map, feel free to use that.

Reply:
If the access is normally happening synchronously, the overhead of locking seems acceptable to me compared to the overhead of a more complex implementation. But unless we have benchmarked it's probably an academic discussion :-)